pczt/src/roles/io_finalizer/mod.rs

```diff
@@ -0,0 +1,105 @@
+use rand_core::OsRng;
+use zcash_primitives::transaction::{
+    components::transparent, sighash::SignableInput, sighash_v5::v5_signature_hash,
+    txid::TxIdDigester,
+};
+
+use crate::Pczt;
+
+use super::signer::pczt_to_tx_data;
+
+pub struct IoFinalizer {
+    pczt: Pczt,
+}
+
+impl IoFinalizer {
+    /// Instantiates the IO Finalizer role with the given PCZT.
+    pub fn new(pczt: Pczt) -> Self {
+        Self { pczt }
+    }
+
+    /// Finalizes the IO of the PCZT.
+    pub fn finalize_io(self) -> Result<Pczt, Error> {
+        let Self { pczt } = self;
+
+        let has_shielded_spends =
+            !(pczt.sapling.spends.is_empty() && pczt.orchard.actions.is_empty());
+        let has_shielded_outputs =
+            !(pczt.sapling.outputs.is_empty() && pczt.orchard.actions.is_empty());
+
+        // We can't build a transaction that has no spends or outputs.
+        // However, we don't attempt to reject an entirely dummy transaction.
+        if pczt.transparent.inputs.is_empty() && !has_shielded_spends {
+            return Err(Error::NoSpends);
+        }
+        if pczt.transparent.outputs.is_empty() && !has_shielded_outputs {
+            return Err(Error::NoOutputs);
+        }
+
+        let Pczt {
+            mut global,
+            transparent,
+            sapling,
+            orchard,
+        } = pczt;
+
+        // After shielded IO finalization, the inputs and outputs cannot be modified
+        // because dummy spends will have been signed.
+        if has_shielded_spends || has_shielded_outputs {
+            global.tx_modifiable &= !0b0000_0011;
+        }
+
+        let transparent = transparent.into_parsed().map_err(Error::TransparentParse)?;
+        let mut sapling = sapling.into_parsed().map_err(Error::SaplingParse)?;
+        let mut orchard = orchard.into_parsed().map_err(Error::OrchardParse)?;
+
+        let tx_data = pczt_to_tx_data(&global, &transparent, &sapling, &orchard)?;
+        let txid_parts = tx_data.digest(TxIdDigester);
+
+        // TODO: Pick sighash based on tx version.
```

Comment:
Is this going to be fixed in this PR? It seems fairly straightforward.

Reply:
It isn't straightforward because with v4 transactions the Sapling proofs need to be created before we can derive a sighash. I'm not actually sure if the current PCZT format even supports v4 transactions (I have not tried creating Sapling proofs before running the IO Finalizer). For now, we will only support v5 transactions, and then we can revisit this when we get v6 (as that will also necessitate a PCZT v2 encoding).

Reply:
Please document this limitation. Also, in that case should we make it impossible for now to create a PCZT with `tx_version < 5`? The current behaviour is just to use the wrong sighash calculation, I think, which seems like a significant footgun.

Reply:
Rather than making it impossible, I'll gate the roles that require sighashing on the version being compatible.